kratos/utilities/variable_utils.h

```diff
@@ -189,6 +190,171 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
         this->CopyModelPartNodalVarToNonHistoricalVar(rVariable, rVariable, rOriginModelPart, rDestinationModelPart, BuffStep);
     }
 
+    /**
+     * @brief Copies flagged nodal historical variable to non-historical variable
+     *
+     * This method copies values in one modelpart's historical nodal container to another modelpart's non-historical container
+     * based on a given flag value is equal to its checking value.
+     *
+     * @tparam TVarType                 Variable type
+     * @param rVariable                 Original model parts variable in historical nodal container where value will be copied from
+     * @param rDestinationVariable      Destination model parts variable where values will be copied to
+     * @param rOriginModelPart          Original modal part where flag and values of nodes will be read
+     * @param rDestinationModelPart     Destination model part where values will be written to
+     * @param rFlag                     Flag to be checked whether to copy data or not
+     * @param CheckValue                Value of flag to be checked
+     * @param BuffStep                  Buffer index, from where the historical nodal values will be read
+     */
+    template <class TVarType>
+    void CopyModelPartFlaggedNodalVarToNonHistoricalVar(
+        const TVarType &rVariable,
+        const TVarType &rDestinationVariable,
+        const ModelPart &rOriginModelPart,
+        ModelPart &rDestinationModelPart,
+        const Flags& rFlag,
+        const bool CheckValue = true,
+        const unsigned int BuffStep = 0)
+    {
+        KRATOS_TRY
+
+        const int n_orig_nodes = rOriginModelPart.NumberOfNodes();
+        const int n_dest_nodes = rDestinationModelPart.NumberOfNodes();
+
+        KRATOS_ERROR_IF_NOT(n_orig_nodes == n_dest_nodes)
+            << "Origin ( " << rOriginModelPart.Name() << " ) and destination ( "
+            << rDestinationModelPart.Name() << " ) model parts have different number of nodes."
+            << "\n\t- Number of origin nodes: " << n_orig_nodes
+            << "\n\t- Number of destination nodes: " << n_dest_nodes << std::endl;
+
+        KRATOS_ERROR_IF_NOT(rOriginModelPart.HasNodalSolutionStepVariable(rVariable))
+            << rVariable.Name() << " not found in nodal solution step data variables list in origin model part ( "
+            << rOriginModelPart.Name() << " ).";
+
+        IndexPartition<int>(n_orig_nodes).for_each([&](int i_node) {
+            const auto p_orig_node = rOriginModelPart.NodesBegin() + i_node;
+            auto p_dest_node = rDestinationModelPart.NodesBegin() + i_node;
+            if (p_orig_node->Is(rFlag) == CheckValue)
+            {
+                p_dest_node->SetValue(
+                    rDestinationVariable,
+                    p_orig_node->FastGetSolutionStepValue(rVariable, BuffStep));
+            }
+        });
```

Comment:
Could you synchronize the variable for mpi  after the values have been transferred? (same for the other function)

As nobody guarantees that all the nodal values (both historical and non historical) are correct while entering this function we could be transferring wrong values.

I think it would be much safer and it should not have a big impact in performance. Thoughts on that @philbucher? (asking you because we faced a very similar situation with the mapper)

Reply:
I also think this makes sense. I added it, waiting for @philbucher thoughts... :)